### PR TITLE
Polish Rubocop configuration

### PR DESCRIPTION
**Pull request by JamesGlover**
- Remove unnecessary options
- Rename Whitelist allowed methods
- Add .rubocop file for useful defaults
- Update rubocop
- Add CountAsOne

Closes #

Changes proposed in this pull request:

*
*
* ...

### CODE REVIEW

File: .rubocop [hunk 1]
```diff
@@ -0,0 +1,5 @@
+--parallel
```
Comment (JamesGlover) [inline .rubocop:1]: Improves performance

File: .rubocop [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+--display-cop-names
+--extra-details
+--display-style-guide
```
Comment (JamesGlover) [inline .rubocop:4]: More verbose error messages (Display cop names is default, but I've included it here for completeness)

File: .rubocop [hunk 3]
```diff
@@ -0,0 +1,5 @@
...
+--display-time
```
Comment (JamesGlover) [inline .rubocop:5]: Maybe this one is just me, but I tend to find reporting of how long steps take useful.

File: .rubocop.yml [hunk 4]
```diff
@@ -117,26 +117,18 @@ Rails/DynamicFindBy:
   Exclude:
     - "app/sample_manifest_excel/**/*"
     - "test/app/sample_manifest_excel/**/*"
-  Whitelist:
+  AllowedMethods:
```
Comment (JamesGlover) [inline .rubocop.yml:120]: Rubocop itself now complains about whitelist, but doesn't check its own files. I'm not quite sure why it doesn't have consistency between AllowedMethods in some cases and IgnoredMethods in others.

File: .rubocop.yml [hunk 5]
```diff
@@ -117,26 +117,18 @@ Rails/DynamicFindBy:
...
     - find_by_sql
     - find_by_id_or_name
     - find_by_id_or_name!
     - find_by_barcode
-    - find_by_or_null
```
Comment (JamesGlover) [inline .rubocop.yml:125]: Not used anymore

File: .rubocop.yml [hunk 6]
```diff
@@ -117,26 +117,18 @@ Rails/DynamicFindBy:
...
 # Method length limits make little sense when you are describing a table schema
 # or a view.
 Metrics/MethodLength:
   Exclude:
     - "db/migrate/*"
-    - "db/migrate_asset_refactor/*"
```
Comment (JamesGlover) [inline .rubocop.yml:131]: Removed this file, so lets remove the exclusion

File: .rubocop.yml [hunk 7]
```diff
@@ -117,26 +117,18 @@ Rails/DynamicFindBy:
...
-Layout/FirstArrayElementIndentation:
-  EnforcedStyle: consistent
```
Comment (JamesGlover) [inline .rubocop.yml:133]: Indentation is now under control of prettier. It configures the layout cops to be compatible, so lets not get in its way. (Even if we seem to be agreeing at the moment)

File: .rubocop.yml [hunk 8]
```diff
@@ -117,26 +117,18 @@ Rails/DynamicFindBy:
...
 Style/ClassAndModuleChildren:
   Enabled: false
-# This should be the rubocop defaults, not sure why its kicking up a fuss.
-Layout/SpaceBeforeBlockBraces:
-  EnforcedStyle: space
-  EnforcedStyleForEmptyBraces: space
```
Comment (JamesGlover) [inline .rubocop.yml:139]: Again, now under control of prettier. (Also, appeared to be a fix for a rubocop bug of some kind)

File: .rubocop.yml [hunk 9]
```diff
@@ -104,39 +100,28 @@ Metrics/BlockLength:
...
     - model
     - step
-    # Less sure that these should be excluded.
-    # Lots of code in them is bad, however we also have things like
-    # expected outputs, which can be large
-    - setup
-    - test
-    - before
-    - should
 # We have a few custom find_by methods which we maintain.
```
Comment (JamesGlover) [inline .rubocop.yml:103]: Counts as one covers my discomfort here.

File: .rubocop.yml [hunk 10]
```diff
@@ -50,12 +48,6 @@ Rails:
 Lint/AmbiguousRegexpLiteral:
   Exclude:
     - features/support/step_definitions/**/*
-# These steps are explicitly designed for debugging. In general usage, they wont
-# be used.
-Lint/Debugger:
-  Exclude:
-    - "features/support/step_definitions/debug_steps.rb"
-    - "features/support/developing.rb"
 # Enforces 1234565789 as 123_456_789: While nice when dealing with actual big
```
Comment (JamesGlover) [inline .rubocop.yml:51]: The latter has been removed, disabled inline for the former

File: .rubocop.yml [hunk 11]
```diff
@@ -17,8 +17,6 @@ RSpec/DescribeClass:
     - "spec/requests/**/*.rb"
     - "spec/api/**/*.rb"
     - "spec/features/**/*.rb"
-    - "spec/lib/warren/*.rb"
-    - "spec/models/billing/*.rb"
 AllCops:
```
Comment (JamesGlover) [inline .rubocop.yml:20]: These files are no more

File: app/helpers/plates_helper.rb [hunk 12]
```diff
@@ -2,12 +2,9 @@ module PlatesHelper # rubocop:todo Style/Documentation
...
   end
 
-  def padded_wells_by_row(plate, overide = nil)
+  def padded_wells_by_row(plate)
     wells = wells_hash(plate)
-    padded_well_name_with_index(plate) do |padded_name, index|
-      index = padded_name == overide ? :overide : index
-      yield(padded_name, *wells[index])
-    end
+    padded_well_name_with_index(plate) { |padded_name, index| yield(padded_name, *wells[index]) }
```
Comment (JamesGlover) [inline app/helpers/plates_helper.rb:7]: And here we go. Bit annoyed that prettier decided to one-line this one, as think the multi line loop was clearer, but :shrug:

File: app/helpers/plates_helper.rb [hunk 13]
```diff
@@ -34,18 +29,17 @@ def well_properties(well)
   end
 
   def padded_well_name_with_index(plate)
-    ('A'...('A'.getbyte(0) + (plate.size / 12)).chr).each_with_index do |row, ri|
-      (1..12).each_with_index do |col, ci|
+    ('A'...('A'.getbyte(0) + (plate.size / 12)).chr).each_with_index do |row, row_index|
```
Comment (JamesGlover) [inline app/helpers/plates_helper.rb:32]: Not rubocop, but spotted these 'concisely' names variables. Was definitely playing code golf when I wrote some of this.

File: app/helpers/plates_helper.rb (hunk 12 above)
Comment (JamesGlover) [reply app/helpers/plates_helper.rb:7]: This follows on from the slack thread https://psd-team.slack.com/archives/C01J32MB747/p1629367326302900

File: .rubocop (hunk 2 above)
Comment (andrewsparkes) [reply .rubocop:4]: yep the more detail the better, will save some googling

File: .rubocop.yml (hunk 7 above)
Comment (andrewsparkes) [reply .rubocop.yml:133]: Yep cool

File: app/controllers/samples_controller.rb [hunk 14]
```diff
@@ -87,8 +83,6 @@ def edit # rubocop:todo Metrics/AbcSize
     end
   end
 
-  # rubocop:enable Metrics/MethodLength
-
   # rubocop:todo Metrics/MethodLength
```
Comment (andrewsparkes) [inline app/controllers/samples_controller.rb:86]: Can that comment go too?
Also, does the methodlength change now factor into AbcSize calculation so some of those can be removed?

File: app/controllers/studies/asset_groups_controller.rb [hunk 15]
```diff
@@ -113,7 +106,6 @@ def search # rubocop:todo Metrics/AbcSize
 
   # rubocop:enable Metrics/MethodLength
```
Comment (andrewsparkes) [inline app/controllers/studies/asset_groups_controller.rb:107]: remove?

File: app/controllers/studies/information_controller.rb [hunk 16]
```diff
@@ -32,8 +31,6 @@ def show # rubocop:todo Metrics/AbcSize
     end
   end
 
-  # rubocop:enable Metrics/MethodLength
-
   # rubocop:todo Metrics/PerceivedComplexity
   # rubocop:todo Metrics/MethodLength
```
Comment (andrewsparkes) [inline app/controllers/studies/information_controller.rb:35]: remove?

File: app/helpers/plates_helper.rb (hunk 12 above)
Comment (andrewsparkes) [reply app/helpers/plates_helper.rb:7]: yes that is not as clear. but not terrible in this instance

File: app/helpers/plates_helper.rb (hunk 13 above)
Comment (andrewsparkes) [reply app/helpers/plates_helper.rb:32]: I prefer your change too, clearer

**Review by andrewsparkes — APPROVED** [review]

**Pull request merged**